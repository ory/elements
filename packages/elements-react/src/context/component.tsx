import { UiNode } from "@ory/client-fetch"
import { PropsWithChildren, createContext, useContext } from "react"
import { OryFlowComponents } from "../types"

type ComponentContextValue = {
  components: OryFlowComponents
  nodeSorter: (a: UiNode, b: UiNode, ctx: { flowType: string }) => number
}

const ComponentContext = createContext<ComponentContextValue>({
  components: null!, // fine because we throw an error if it's not provided
  nodeSorter: () => 0,
})

export function useComponents() {
  const ctx = useContext(ComponentContext)
  if (!ctx) {
    throw new Error("useComponents must be used within a ComponentProvider")
  }
  return ctx.components
}

export function useNodeSorter() {
  const ctx = useContext(ComponentContext)
  if (!ctx) {
    throw new Error("useComponents must be used within a ComponentProvider")
  }
  return ctx.nodeSorter
}

const defaultGroupOrder = [
  "oidc",
  "identifier_first",
  "default",
  "profile",
  "password",
  "passkey",
  "code",
  "webauthn",
]

function defaultNodeSorter(
  a: UiNode,
  b: UiNode,
  // ctx: { flowType: string },
): number {
  const aGroupWeight = defaultGroupOrder.indexOf(a.group) ?? 999
  const bGroupWeight = defaultGroupOrder.indexOf(b.group) ?? 999

  return aGroupWeight - bGroupWeight
}

type ComponentProviderProps = {
  components: OryFlowComponents
  nodeSorter?: (a: UiNode, b: UiNode, ctx: { flowType: string }) => number
}

export function OryComponentProvider({
  children,
  components,
  nodeSorter = defaultNodeSorter,
}: PropsWithChildren<ComponentProviderProps>) {
  return (
    <ComponentContext.Provider
      value={{
        components,
        nodeSorter,
      }}
    >
      {children}
    </ComponentContext.Provider>
  )
}
