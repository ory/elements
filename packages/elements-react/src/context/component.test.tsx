// Copyright Â© 2025 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { UiNode, isUiNodeInputAttributes } from "@ory/client-fetch"
import { defaultNodeSorter } from "./component"

import captchaFormData from "../../.stub-responses/login/1fa/identifier_first/captcha/initial-form.json"
import passwordFormData from "../../.stub-responses/login/1fa/unified/password/initial-form.json"
import registrationData from "../../.stub-responses/registration/one-step/captcha/initial-form.json"
import registrationDataTwoStep from "../../.stub-responses/registration/two-step/captcha/initial-form.json"

describe("defaultNodeSorter", () => {
  it("correctly sorts nodes from captcha form response", () => {
    // Load test data from stub file
    const nodes: UiNode[] = captchaFormData.ui.nodes as UiNode[]

    // Sort the nodes
    const sortedNodes = [...nodes].sort(defaultNodeSorter)

    // Create a simplified representation for the snapshot
    const nodeOrder = sortedNodes.map((node) => ({
      group: node.group,
      type: node.attributes.node_type,
      inputType: isUiNodeInputAttributes(node.attributes)
        ? node.attributes.type
        : undefined,
      name: isUiNodeInputAttributes(node.attributes)
        ? node.attributes.name
        : undefined,
    }))

    // Snapshot test for consistent ordering
    expect(nodeOrder).toMatchSnapshot("captcha form node order")

    // Find captcha nodes and submit buttons to verify ordering
    const captchaNodes = sortedNodes.filter((node) => node.group === "captcha")
    const submitButtons = sortedNodes.filter(
      (node) =>
        isUiNodeInputAttributes(node.attributes) &&
        node.attributes.type === "submit",
    )

    if (captchaNodes.length > 0 && submitButtons.length > 0) {
      const lastCaptchaIndex = sortedNodes.indexOf(
        captchaNodes[captchaNodes.length - 1],
      )
      const firstSubmitIndex = sortedNodes.indexOf(submitButtons[0])

      // This test should now pass with the fixed node sorter
      expect(lastCaptchaIndex).toBeLessThan(firstSubmitIndex)
    }
  })

  it("correctly sorts nodes from unified password form response", () => {
    // Load test data from stub file
    const nodes: UiNode[] = passwordFormData.ui.nodes as UiNode[]

    // Sort the nodes
    const sortedNodes = [...nodes].sort(defaultNodeSorter)

    // Create a simplified representation for the snapshot
    const nodeOrder = sortedNodes.map((node) => ({
      group: node.group,
      type: node.attributes.node_type,
      inputType: isUiNodeInputAttributes(node.attributes)
        ? node.attributes.type
        : undefined,
      name: isUiNodeInputAttributes(node.attributes)
        ? node.attributes.name
        : undefined,
    }))

    // Snapshot test for consistent ordering
    expect(nodeOrder).toMatchSnapshot("unified password form node order")

    // Verify captcha nodes come before submit buttons
    const captchaNodes = sortedNodes.filter((node) => node.group === "captcha")
    const submitButtons = sortedNodes.filter(
      (node) =>
        isUiNodeInputAttributes(node.attributes) &&
        node.attributes.type === "submit",
    )

    if (captchaNodes.length > 0 && submitButtons.length > 0) {
      const lastCaptchaIndex = sortedNodes.indexOf(
        captchaNodes[captchaNodes.length - 1],
      )
      const firstSubmitIndex = sortedNodes.indexOf(submitButtons[0])

      // This test should now pass with the fixed sorter
      expect(lastCaptchaIndex).toBeLessThan(firstSubmitIndex)
    }
  })

  it("correctly sorts nodes from one-step registration with captcha", () => {
    // Load test data from stub file
    const nodes: UiNode[] = registrationData.ui.nodes as UiNode[]

    // Sort the nodes
    const sortedNodes = [...nodes].sort(defaultNodeSorter)

    // Create a simplified representation for the snapshot
    const nodeOrder = sortedNodes.map((node) => ({
      group: node.group,
      type: node.attributes.node_type,
      inputType: isUiNodeInputAttributes(node.attributes)
        ? node.attributes.type
        : undefined,
      name: isUiNodeInputAttributes(node.attributes)
        ? node.attributes.name
        : undefined,
    }))

    // Snapshot test for consistent ordering
    expect(nodeOrder).toMatchSnapshot(
      "one-step registration with captcha node order",
    )

    // Verify captcha nodes come before submit buttons
    const captchaNodes = sortedNodes.filter((node) => node.group === "captcha")
    const submitButtons = sortedNodes.filter(
      (node) =>
        isUiNodeInputAttributes(node.attributes) &&
        node.attributes.type === "submit",
    )

    if (captchaNodes.length > 0 && submitButtons.length > 0) {
      const lastCaptchaIndex = sortedNodes.indexOf(
        captchaNodes[captchaNodes.length - 1],
      )
      const firstSubmitIndex = sortedNodes.indexOf(submitButtons[0])

      expect(lastCaptchaIndex).toBeLessThan(firstSubmitIndex)
    }
  })

  it("correctly sorts nodes from two-step registration with captcha", () => {
    // Load test data from stub file
    const nodes: UiNode[] = registrationDataTwoStep.ui.nodes as UiNode[]

    // Sort the nodes
    const sortedNodes = [...nodes].sort(defaultNodeSorter)

    // Create a simplified representation for the snapshot
    const nodeOrder = sortedNodes.map((node) => ({
      group: node.group,
      type: node.attributes.node_type,
      inputType: isUiNodeInputAttributes(node.attributes)
        ? node.attributes.type
        : undefined,
      name: isUiNodeInputAttributes(node.attributes)
        ? node.attributes.name
        : undefined,
    }))

    // Snapshot test for consistent ordering
    expect(nodeOrder).toMatchSnapshot(
      "two-step registration with captcha node order",
    )

    // Verify captcha nodes come before submit buttons
    const captchaNodes = sortedNodes.filter((node) => node.group === "captcha")
    const submitButtons = sortedNodes.filter(
      (node) =>
        isUiNodeInputAttributes(node.attributes) &&
        node.attributes.type === "submit",
    )

    if (captchaNodes.length > 0 && submitButtons.length > 0) {
      const lastCaptchaIndex = sortedNodes.indexOf(
        captchaNodes[captchaNodes.length - 1],
      )
      const firstSubmitIndex = sortedNodes.indexOf(submitButtons[0])

      expect(lastCaptchaIndex).toBeLessThan(firstSubmitIndex)
    }
  })
})
