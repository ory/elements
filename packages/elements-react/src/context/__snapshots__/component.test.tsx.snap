// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultNodeSorter correctly sorts nodes from captcha form response: captcha form node order 1`] = `
[
  {
    "group": "identifier_first",
    "inputType": "text",
    "name": "identifier",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "hidden",
    "name": "csrf_token",
    "type": "input",
  },
  {
    "group": "captcha",
    "inputType": "hidden",
    "name": "transient_payload.captcha_turnstile_response",
    "type": "input",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "script",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "div",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "script",
  },
  {
    "group": "captcha",
    "inputType": "hidden",
    "name": "captcha_turnstile_options",
    "type": "input",
  },
  {
    "group": "identifier_first",
    "inputType": "submit",
    "name": "method",
    "type": "input",
  },
  {
    "group": "passkey",
    "inputType": "hidden",
    "name": "passkey_login",
    "type": "input",
  },
  {
    "group": "passkey",
    "inputType": "hidden",
    "name": "passkey_challenge",
    "type": "input",
  },
  {
    "group": "webauthn",
    "inputType": undefined,
    "name": undefined,
    "type": "script",
  },
]
`;

exports[`defaultNodeSorter correctly sorts nodes from one-step registration with captcha: one-step registration with captcha node order 1`] = `
[
  {
    "group": "default",
    "inputType": "hidden",
    "name": "csrf_token",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "email",
    "name": "traits.email",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "checkbox",
    "name": "traits.tos",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "text",
    "name": "traits.phone",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "text",
    "name": "traits.nested.name",
    "type": "input",
  },
  {
    "group": "password",
    "inputType": "password",
    "name": "password",
    "type": "input",
  },
  {
    "group": "captcha",
    "inputType": "hidden",
    "name": "transient_payload.captcha_turnstile_response",
    "type": "input",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "script",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "div",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "script",
  },
  {
    "group": "captcha",
    "inputType": "hidden",
    "name": "captcha_turnstile_options",
    "type": "input",
  },
  {
    "group": "password",
    "inputType": "submit",
    "name": "method",
    "type": "input",
  },
  {
    "group": "passkey",
    "inputType": "button",
    "name": "passkey_register_trigger",
    "type": "input",
  },
  {
    "group": "passkey",
    "inputType": "hidden",
    "name": "passkey_register",
    "type": "input",
  },
  {
    "group": "passkey",
    "inputType": "hidden",
    "name": "passkey_create_data",
    "type": "input",
  },
  {
    "group": "webauthn",
    "inputType": undefined,
    "name": undefined,
    "type": "script",
  },
]
`;

exports[`defaultNodeSorter correctly sorts nodes from two-step registration with captcha: two-step registration with captcha node order 1`] = `
[
  {
    "group": "default",
    "inputType": "hidden",
    "name": "csrf_token",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "email",
    "name": "traits.email",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "checkbox",
    "name": "traits.tos",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "text",
    "name": "traits.phone",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "text",
    "name": "traits.nested.name",
    "type": "input",
  },
  {
    "group": "captcha",
    "inputType": "hidden",
    "name": "transient_payload.captcha_turnstile_response",
    "type": "input",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "script",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "div",
  },
  {
    "group": "captcha",
    "inputType": undefined,
    "name": undefined,
    "type": "script",
  },
  {
    "group": "captcha",
    "inputType": "hidden",
    "name": "captcha_turnstile_options",
    "type": "input",
  },
  {
    "group": "profile",
    "inputType": "submit",
    "name": "method",
    "type": "input",
  },
]
`;

exports[`defaultNodeSorter correctly sorts nodes from unified password form response: unified password form node order 1`] = `
[
  {
    "group": "default",
    "inputType": "hidden",
    "name": "csrf_token",
    "type": "input",
  },
  {
    "group": "default",
    "inputType": "text",
    "name": "identifier",
    "type": "input",
  },
  {
    "group": "password",
    "inputType": "password",
    "name": "password",
    "type": "input",
  },
  {
    "group": "password",
    "inputType": "submit",
    "name": "method",
    "type": "input",
  },
]
`;
